### PR TITLE
30921 - fix GraphQL response not compliant with UI behavior

### LINKED ISSUE #30921: GraphQL response not compliant with UI behavior
### Preconditions (*)

1. Magento 2.4.1 CE

### Steps to reproduce (*)

The bug described below affects Bundle and Configurable Product. Might affect other kinds of product grouping but I haven't tested.

1. Create or edit a **Bundle** or **Configurable** Product
2. Add products to bundle or configuration
3. Save the bundle/configuration.
4. Open one of the products previously added to the bundle/configuration
5. Add some customizable options and save the product
6. Open the bundle/configuration in the store Frontend. The item with customizable options is not displayed. **This is normal behavior**.
7. Query the bundle product via graphQL (example query below). The item with customizable options is part of the response. **The response should reflect the UI behavior**.

```graphql
{
  products(filter: {url_key: {eq: "test-bundle"}}) {
    items {
      sku
      name
      ... on BundleProduct {
        items {
          option_id
          title
          options {
            id
            label
          }
        }
      }
    }
  }
}
```

```graphql
{
  products(filter: {url_key: {eq: "chaz-kangeroo-hoodie"}}) {
    items {
      sku
      name
      ... on ConfigurableProduct {
        variants {
          product {
            sku
            name
          }
        }
      }
    }
  }
}
```

### Expected result (*)

1. The graphQL response for items with customizable options in a Bundle or Configurable Product should have the same behavior as the UI Frontend

### Actual result (*)

1. Bundle in the Magento UI (Joust Duffle Bag has Customizable options):
image

2. GraphQL Bundle Product Response:
image

3. Configurable Product int the UI (the XS-Black variation has customizable options)
image

4. GraphQL Configurable Product Response:
image

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [x] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by korovitskyi**
### Description (*)
Fixed the issue `GraphQL response not compliant with UI behavior`

### Fixed Issues (if relevant)

1. Fixes magento/magento2#30921

### Manual testing scenarios (*)
1. Create or edit a Bundle or Configurable Product
2. Add products to bundle or configuration
3. Save the bundle/configuration.
4. Open one of the products previously added to the bundle/configuration
5. Add some customizable options and save the product
6. Open the bundle/configuration in the store Frontend. The item with customizable options is not displayed. This is normal behavior.
7. Query the bundle product via graphQL (example query below). The item with customizable options is part of the response. The response should reflect the UI behavior.
```
{
  products(filter: {url_key: {eq: "test-bundle"}}) {
    items {
      sku
      name
      ... on BundleProduct {
        items {
          option_id
          title
          options {
            id
            label
          }
        }
      }
    }
  }
}
```
```
{
  products(filter: {url_key: {eq: "chaz-kangeroo-hoodie"}}) {
    items {
      sku
      name
      ... on ConfigurableProduct {
        variants {
          product {
            sku
            name
          }
        }
      }
    }
  }
}

```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by rogyar — CHANGES_REQUESTED** [review]
Hi @korovitskyi. Thank you for your collaboration. Please, sign [Adobe CLA](http://opensource.adobe.com/cla.html) so we can proceed with processing your PR. According to the [definition of done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html), all the changes should be covered by autotests. Could I ask you to cover your changes with an API-functional test? 

You may try to extend the existing tests.

https://github.com/magento/magento2/blob/252e59659488dba0090f182a5eed16eac5db5bed/dev/tests/api-functional/testsuite/Magento/GraphQl/Bundle/BundleProductMultipleOptionsTest.php#L17

https://github.com/magento/magento2/blob/42ac2e8fb2c9438ff656e3962f7395cb601fadcd/dev/tests/api-functional/testsuite/Magento/GraphQl/ConfigurableProduct/ConfigurableProductOptionsTest.php#L15

Thank you

File: dev/tests/integration/testsuite/Magento/Bundle/_files/product_with_multiple_options_product_with_custom_options.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                $option->setProductLinks($links);
+                $options[] = $option;
+            }
+        }
+    }
+    $extension = $product->getExtensionAttributes();
+    $extension->setBundleProductOptions($options);
+    $product->setExtensionAttributes($extension);
+}
+$product->save();
```
Comment (rogyar) [inline dev/tests/integration/testsuite/Magento/Bundle/_files/product_with_multiple_options_product_with_custom_options.php]: Please, don't use deprecated methods in your fixtures. Use the repository or the resource model for saving the product. 

Thanks

**Review by rogyar — CHANGES_REQUESTED** [review]

File: dev/tests/integration/testsuite/Magento/Bundle/_files/product_with_multiple_options_product_with_custom_options.php (hunk 1 above)
Comment (korovitskyi) [reply dev/tests/integration/testsuite/Magento/Bundle/_files/product_with_multiple_options_product_with_custom_options.php]: @rogyar ,thanks for review.
This case was fixed.

**Comment by rogyar** [comment]
@magento run Static Tests, WebAPI Tests

**Review by rogyar — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-8540](https://jira.corp.magento.com/browse/ENGCOM-8540) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Hi @korovitskyi & @rogyar Thanks for your contribution and collaboration!

Tried to reproduce the same issue in 2.4-develop Magento code, and seems to be the issue has fixed already.
Below is the update on what I executed.

1. Created a bundle product as shown in the below screenshot.
image

2. Executed the GraphQL query with the request body:
`{
  products(filter: {url_key: {eq: "mksbundle1"}}) {
    items {
      sku
      name
      ... on BundleProduct {
        items {
          option_id
          title
          options {
            id
            label
          }
        }
      }
    }
  }
}`

5. Validated that the response matching the UI behaviour, and the response received is:
`{
    "data": {
        "products": {
            "items": [
                {
                    "sku": "mksbundle1",
                    "name": "mksbundle1",
                    "items": [
                        {
                            "option_id": 5,
                            "title": "optionchkbx",
                            "options": [
                                {
                                    "id": 10,
                                    "label": "Affirm Water Bottle "
                                },
                                {
                                    "id": 11,
                                    "label": "Go-Get'r Pushup Grips"
                                }
                            ]
                        },
                        {
                            "option_id": 6,
                            "title": "optionrd",
                            "options": [
                                {
                                    "id": 12,
                                    "label": "Pursuit Lumaflex&trade; Tone Band"
                                },
                                {
                                    "id": 13,
                                    "label": "Zing Jump Rope"
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    }
}`
Noticed it is same as UI behaviour.

4. Used a configurable product and set one option _out of stock_ as shown in the below screenshot.
image

5. Executed the GraphQL query with the request body:
`{
  products(filter: {url_key: {eq: "chaz-kangeroo-hoodie"}}) {
    items {
      sku
      name
      ... on ConfigurableProduct {
        variants {
          product {
            sku
            name
          }
        }
      }
    }
  }
}`
6. Validated that the response matching the UI behaviour, and the response received is:
`{
    "data": {
        "products": {
            "items": [
                {
                    "sku": "MH01",
                    "name": "Chaz Kangeroo Hoodie",
                    "variants": [
                        {
                            "product": {
                                "sku": "MH01-XS-Gray",
                                "name": "Chaz Kangeroo Hoodie-XS-Gray"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-XS-Orange",
                                "name": "Chaz Kangeroo Hoodie-XS-Orange"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-S-Black",
                                "name": "Chaz Kangeroo Hoodie-S-Black"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-S-Gray",
                                "name": "Chaz Kangeroo Hoodie-S-Gray"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-S-Orange",
                                "name": "Chaz Kangeroo Hoodie-S-Orange"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-M-Black",
                                "name": "Chaz Kangeroo Hoodie-M-Black"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-M-Gray",
                                "name": "Chaz Kangeroo Hoodie-M-Gray"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-M-Orange",
                                "name": "Chaz Kangeroo Hoodie-M-Orange"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-L-Black",
                                "name": "Chaz Kangeroo Hoodie-L-Black"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-L-Gray",
                                "name": "Chaz Kangeroo Hoodie-L-Gray"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-L-Orange",
                                "name": "Chaz Kangeroo Hoodie-L-Orange"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-XL-Black",
                                "name": "Chaz Kangeroo Hoodie-XL-Black"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-XL-Gray",
                                "name": "Chaz Kangeroo Hoodie-XL-Gray"
                            }
                        },
                        {
                            "product": {
                                "sku": "MH01-XL-Orange",
                                "name": "Chaz Kangeroo Hoodie-XL-Orange"
                            }
                        }
                    ]
                }
            ]
        }
    }
}`
Noticed it is same as UI behaviour again.

The issue must have fixed already with other fixes in the past, can you please check and let me know if we are expecting in other fix using this PR. @korovitskyi

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Require an confirmation on [above-comment](https://github.com/magento/magento2/pull/31072#issuecomment-1102479930), and if it is still need to be considered, then require merge conflicts to be resolved along with detailed response on above-comment. Hence moving it to _Changes Requested_. Informed to @korovitskyi over the slack too!
Thanks in advance!

**Comment by engcom-Alfa** [comment]
Hi @korovitskyi / @rogyar 
Apart from above-commented scenarios aside, re-retried as mentioned here in the [issue-description](https://github.com/magento/magento2/issues/30921#issue-742565541) as well, and below is the same procedure.
1. Created 3 simple products.
2. Created a bundle products adding above created simple products as bundle options, here is the screenshot of it
    image
3. Edited the simple product 1 and added a customizable option to it, as shown in below screenshot.
    image
4. Tried to save the edited simple product 1. Since we have already added the simple product to the bundle options, the Magento doesn't allow us to add customizable options into the simple product. We get error as shown in below screenshot.
    image

So, we can not proceed ahead with the exact steps that is defined in the [issue-description](https://github.com/magento/magento2/issues/30921#issue-742565541). If we are missing anything as per above given steps here or above las commented steps, kindly correct us on this to reproduce the issue.

For time being, I am closing this PR ticket. Once after getting your reply on this anytime, we will re-open the ticket again and continue working on this.

Thanks for your contribution :)

**Pull request closed without merging**